src/www/scripts/app.js

```diff
@@ -622,26 +631,46 @@ module.exports = (function() {
 
             setupDirectoryLocations();
 
+            var tokenStore = requirePin ? new TokenStore(secureStore) : new TokenStore(localStore);
+
             var shouldDownloadFn = function(config) {
                 return config.downloadResources || enableOffline;
             };
 
-            if (requirePin) {
-                var tokenStore = new TokenStore(secureStore);
+            var cleanUpTokensFn = function() {
+                return BPromise
+                    .all([tokenStore.remove(), pin.remove()])
+                    .catch(function() {
+                        console.info("Could not clean tokenStore and PIN; maybe they were already removed.");
+                    });
+            };
 
-                BPromise.all([ tokenStore.get(), pin.get() ]).spread(function(token, storedPin) {
-                    if (token && storedPin) {
-                        pinView.verify(syncAndStartup);
-                    } else {
-                        BPromise.all([ tokenStore.remove(), pin.remove() ]).then(syncAndStartup);
-                    }
-                });
-            } else {
-                syncAndStartup();
-            }
+            new BPromise(function(resolve, reject) {
+                if (credentialsProvided(username, password)) {
+                    cleanUpTokensFn()
+                        .then(function() {
+                            return createSessionWithCredentials(appUrl, username, password);
+                        })
+                        .then(resolve, reject);
+                } else if (requirePin) {
+                    BPromise
+                        .all([tokenStore.get(), pin.get()])
+                        .spread(function(token, storedPin) {
+                            if (token && storedPin) {
+                                pinView.verify(resolve);
+                            } else {
+                                return cleanUpTokensFn()
+                                    .then(resolve);
+                            }
+                        });
+                } else {
+                    resolve();
+                }
+            }).catch(function(e) {
+                handleError(e);
+            }).finally(syncAndStartup);
```

Comment:
I found it extremely difficult to read this piece of code, mostly because the cyclomatic and/or continuation complexity seems pretty high. I recommend we set some time apart to refactor this.